zerver/models.py

```diff
@@ -1801,10 +1801,42 @@ class Subscription(models.Model):
     color = models.CharField(max_length=10, default=DEFAULT_STREAM_COLOR)  # type: str
     pin_to_top = models.BooleanField(default=False)  # type: bool
 
-    desktop_notifications = models.BooleanField(default=True)  # type: bool
-    audible_notifications = models.BooleanField(default=True)  # type: bool
-    push_notifications = models.BooleanField(default=False)  # type: bool
-    email_notifications = models.BooleanField(default=False)  # type: bool
+    _desktop_notifications = models.BooleanField(default=None, null=True)  # type: Optional[bool]
+    _audible_notifications = models.BooleanField(default=None, null=True)  # type: Optional[bool]
+    _push_notifications = models.BooleanField(default=None, null=True)  # type: Optional[bool]
+    _email_notifications = models.BooleanField(default=None, null=True)  # type: Optional[bool]
+
+    @property
+    def desktop_notifications(self) -> bool:
```

Comment:
This adds a lot of duplicate code but I think is better than using a dictionary based approach because first it is more readable and second that mypy gives a nice error if someone accidentally sets like `desktop_notifications` instead of `_desktop_notifications`. What do you think?

Reply:
This direction of interesting, though I think we might be better off with just keeping the existing names.  Part of the reason is that for desktop and audible notifications, it's presumably the frontends that will be making the determination (and so we'll want to send them `None`), and also I think this strategy might interfere with our `property_types` framework.  In any case I think we only need to do lookups in roughly one place.

(Looks like we need to use NullBooleanField.  )